server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -156,34 +166,97 @@ public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listene
         return true;
     }
 
+    /**
+     * Add a listener for refreshes, calling it immediately if the location is already visible. If this runs out of listener slots then it
+     * forces a refresh and calls the listener immediately as well. The checkpoint cannot be greater than the processed local checkpoint.
+     *
+     * @param checkpoint the seqNo checkpoint to listen for
+     * @param listener for the refresh.
+     * @return did we call the listener (true) or register the listener to call later (false)?
+     */
+    public boolean addOrNotify(long checkpoint, ActionListener<Void> listener) {
+        assert checkpoint >= SequenceNumbers.NO_OPS_PERFORMED;
+        if (checkpoint <= lastRefreshedCheckpoint) {
+            listener.onResponse(null);
+            return true;
+        }
+        long maxIssuedSequenceNumber = maxIssuedSeqNoSupplier.getAsLong();
+        if (checkpoint > maxIssuedSequenceNumber) {
+            IllegalArgumentException e = new IllegalArgumentException("Cannot wait for unissued seqNo checkpoint [wait_for_checkpoint="
+                + checkpoint + ", max_issued_seqNo=" + maxIssuedSequenceNumber + "]");
+            listener.onFailure(e);
+            return true;
+        }
+
+        synchronized (this) {
+            if (closed) {
+                listener.onFailure(new IllegalStateException("can't wait for refresh on a closed index"));
+                return true;
+            }
+            List<Tuple<Long, ActionListener<Void>>> listeners = checkpointRefreshListeners;
+            final int maxRefreshes = getMaxRefreshListeners.getAsInt();
+            if (refreshForcers == 0 && roomForListener(maxRefreshes, locationRefreshListeners, listeners)) {
+                addCheckpointListener(checkpoint, listener, listeners);
+                return false;
+            }
+        }
+        // No free slot so force a refresh and call the listener in this thread
+        forceRefresh.run();
```

Comment:
we could also raise an exception at this point ? If there is a replica we'll try on another node and otherwise the caller will get a 429 try again later ? That would avoid performing the refresh on the network thread.

Reply:
I looked into this and it looks like it would be a lot of work? RefreshListeners currently has functionality around forcing refreshes when listeners are added. `RefreshListeners #forceRefreshes()`. And I am hesitant to wade into change that behavior in this PR?

Reply:
I need to look into this a bit more before making this change.

Reply:
And you're suggesting we just allow unlimited refresh listeners? That is another reason we force the refresh.